Generate new library version

brettmortensen opened the pull request:
First cut at generating a new version of the wrapper library using
code-gen tools.

The generated stock file structure does not look right, I'm thinking of separating the models into a `models` directory and api files into a `client` directory.

Any feedback is appreciated 👍 

@film42 @minond

brettmortensen committed:
Generate new library version

First cut at generating a new version of the wrapper library using
code-gen tools.

minond commented on api_connect_widget.go:
```diff
@@ -0,0 +1,143 @@
+ * API version: 0.1
+ */
+
+package atrium-go
```
This is a syntax error.

minond commented on api_connect_widget.go:
```diff
@@ -0,0 +1,143 @@
+	}
+	if ctx != nil {
+		// API Key Authentication
+		if auth, ok := ctx.Value(ContextAPIKey).(APIKey); ok {
```
Looks to be using the same key (`ContextAPIKey`) for both the API key and the Client ID values.

minond commented on client.go:
```diff
@@ -0,0 +1,484 @@
+				return err
+			}
+			return nil
+		} else if strings.Contains(contentType, "json") {
```
Oh nice, this should take care of the decoding issue.

minond reviewed (changes requested):
At first glance things look OK to me. Just the package name issue (can't have a hyphen and it should just be "atrium" IMO) and the ClientID key issue, that's all I noticed. If you need a hand running some of this code LMK.

minond commented:
@brettmortensen what were the issues we found the first time we tried this? Missing client IDs, weird request/response types, json decoding failing by default, anything else?

film42 commented on README.md:
````diff
@@ -1,53 +1,96 @@
+## Installation
+Put the package under your project folder and add the following in import:
+```golang
+import "./atrium-go"
````
This is wrong. You'll be importing `github.com/mxenabled/atrium-go`.

film42 commented on LICENSE:
```diff
@@ -1,21 +0,0 @@
-The MIT License (MIT)
```
Why are we removing the MIT license? This code is open source and free to use/ change.

brettmortensen committed:
Fix issues

Renames package name without hyphen to avoid syntax error. Cleans up
authentication by overriding methods. Corrects relative import with
github repo. Adds back in MIT license.

brettmortensen commented:
@minond I believe you listed all of the issues we found above. I think they have all been resolved now besides the generic request/response types.

minond commented:
@brettmortensen just messaged you some solutions for the type issues we are running into. Let me know if they help.

brettmortensen committed:
Use specific types over generic interfaces

brettmortensen commented:
@minond Thanks, I've updated the library with the solution you suggested. It appears to solve the issue we were seeing with the interfaces. Let me know what you think 👍

minond commented:
@brettmortensen that's awesome. I took a quick peek at the user create payload body and it looks like it's still using an `interface{}`. Did all of the definitions get updated or just some of them?

minond commented:
I take back my last comment. I'll take a look later today when I free up a bit more!

brettmortensen committed:
Remove additional branding

brettmortensen committed:
Add missing 'type' field to challenge

minond commented:
Looking good to me. I like the new constructors you created, really cleans things up. Digging around and the types look much cleaner as well -- happy not to see any `interface{}`'s. The only thing I don't like is the whole "attributes" convention that we have right now. For example, `User` is defined as

```go
type User struct {
	User *UserAttributes `json:"user,omitempty"`
}
```

so creating one requires doing

```go
user := atrium.User{
	User: &atrium.UserAttributes{
		Guid: "123",
	},
}
```

but really I just want to be able to

```go
user := atrium.User{
	Guid: "123",
}
```

I'm not sure if there's a reason we're having to do this right now, but if we could do away with a "ModelAttribute" and just work with a "Model" that'd be nice. I think we should clean this up before releasing this.

The other thing is more a nitpick than anything else, and that's that we should follow Go's naming convention for things like "URL", "GUID", "ID", etc. Running a linter on this project (https://github.com/golang/lint) I'm seeing lots of warnings about "userGuid" needing to be "userGUID", "MediumLogoUrl" needing to be "MediumLogoURL", and so on. If you do a `go get -u golang.org/x/lint/golint` and then run `golint` at the root of this project you can see the full list yourself. You'll see warnings about missing comments, using underscores, and a few other naming conventions not being followed. I say we focus on properly naming model fields (eg `User.Guid -> User.GUID`) and API objects (eg `client.UsersApi -> client.UsersAPI`) right now and worry less about the rest for now. I think those names are more important since users will actually be using those.

brettmortensen committed:
Properly name model fields and API Objects

brettmortensen commented:
@minond Thanks, great feedback. 

The naming of the model fields and API objects has been updated.

Good point about the "attributes" pattern I followed. When setting this up originally, I followed some swagger examples that used the same pattern. I've been searching/trying to get rid of the nested "attributes" model, but it doesn't appear there is an easy way to do that in the generators. Any time there are nested objects, it forces them to be individual objects 👎

brettmortensen committed:
Refactor models with request/response pattern

brettmortensen commented:
@minond I just updated the models with the "request"/"response" pattern we discussed yesterday offline. It should clear up the confusion with the previous "attributes" pattern.

brettmortensen committed:
Add example snippet to readme

brettmortensen committed:
Remove php line

brettmortensen committed:
Use single api client for all resources

brettmortensen committed:
Update API wrapper and  examples

brettmortensen committed:
Remove travis.yml from previous version

brettmortensen committed:
Update version to 2.0.0

brettmortensen committed:
Setup travis CI

film42 committed:
Bump go version to 1.11

film42 committed:
Add go mod file

film42 committed:
Remove style check for now

film42 committed:
Enable style checks in CI

film42 committed:
Run gofmt

film42 commented:
:shipit: 

Let's make sure to tag this as `v2.0.0`.
